datafusion/datasource-csv/src/file_format.rs

```diff
@@ -560,21 +572,27 @@ impl CsvFormat {
                     })
                     .unzip();
             } else {
-                if fields.len() != column_type_possibilities.len() {
-                    return exec_err!(
-                            "Encountered unequal lengths between records on CSV file whilst inferring schema. \
-                             Expected {} fields, found {} fields at record {}",
-                            column_type_possibilities.len(),
-                            fields.len(),
-                            record_number + 1
-                        );
+                // Handle files with different numbers of columns by extending the schema
+                if fields.len() > column_type_possibilities.len() {
+                    // New columns found - extend our tracking structures
+                    for field in fields.iter().skip(column_type_possibilities.len()) {
+                        column_names.push(field.name().clone());
+                        let mut possibilities = HashSet::new();
+                        if records_read > 0 {
+                            possibilities.insert(field.data_type().clone());
+                        }
+                        column_type_possibilities.push(possibilities);
+                    }
                 }
 
-                column_type_possibilities.iter_mut().zip(&fields).for_each(
-                    |(possibilities, field)| {
-                        possibilities.insert(field.data_type().clone());
-                    },
-                );
+                // Update type possibilities for columns that exist in this file
+                // Only process fields that exist in both the current file and our tracking structures
+                for (field_idx, field) in fields.iter().enumerate() {
+                    if field_idx < column_type_possibilities.len() {
+                        column_type_possibilities[field_idx]
+                            .insert(field.data_type().clone());
+                    }
+                }
```

Comment:
It might be less confusing if we keep the original code:

```rust
column_type_possibilities.iter_mut().zip(&fields).for_each(
    |(possibilities, field)| {
        possibilities.insert(field.data_type().clone());
    },
);
```

But move it before the `Handle files with different numbers of columns by extending the schema` check (since zip will terminate on the shortest iterator), so we first extend possibilities for existing columns, then afterwards add the new columns.

Another way could be to do both in a single iteration; something like

```
for idx, field in enumerate(fields)
    if idx < len(possibilities)
        add_possibility
    else
        add_new_field
```

Thoughts?